### PR TITLE
[PVR] Refactor PVR jobs. Get rid of PVRJobs.(h|cpp).

### PR DESCRIPTION
**Pull request by ksooo**
Refactoring of PVR jobs. Basic ideas: Use lambda jobs to reduce boilerplate code and better encapsulation by reducing visibility of the job implementations. 

Runtime-tested on macOS and Android, latest Kodi master.

### CODE REVIEW

File: xbmc/pvr/PVREventlogJob.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+{
+  for (const auto& event : m_events)
+  {
+    if (event.m_bNotifyUser)
+      CGUIDialogKaiToast::QueueNotification(
+        event.m_bError ? CGUIDialogKaiToast::Error : CGUIDialogKaiToast::Info, event.m_label.c_str(), event.m_msg, 5000, true);
+
+    // Write event log entry.
+    CServiceBroker::GetEventLog().Add(
+      EventPtr(new CNotificationEvent(event.m_label, event.m_msg, event.m_icon, event.m_bError ? EventLevel::Error : EventLevel::Information)));
```
Comment (phunkyfish) [inline xbmc/pvr/PVREventlogJob.cpp]: make_shared?

**Review by phunkyfish — CHANGES_REQUESTED** [review]
Just a couple of things

File: xbmc/pvr/CMakeLists.txt [hunk 2]
```diff
@@ -7,7 +7,7 @@ set(SOURCES PVRActionListener.cpp
             PVRGUIActions.cpp
             PVRItem.cpp
             PVRChannelNumberInputHandler.cpp
-            PVRJobs.cpp
+            PVREventlogJob.cpp
```
[the remark this answers is not in the document]
Comment (ksooo) [reply xbmc/pvr/CMakeLists.txt]: Yep, good point. Will change that.

File: xbmc/pvr/PVREventlogJob.cpp (hunk 1 above)
Comment (ksooo) [reply xbmc/pvr/PVREventlogJob.cpp]: sure.

File: xbmc/pvr/PVREventlogJob.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+      std::string m_label;
+      std::string m_msg;
+      std::string m_icon;
+
+      Event(bool bNotifyUser, bool bError, const std::string& label, const std::string& msg, const std::string& icon)
+      : m_bNotifyUser(bNotifyUser), m_bError(bError), m_label(label), m_msg(msg), m_icon(icon) {}
+    };
+
+    std::vector<Event> m_events;
+  };
```
Comment (Rechi) [inline xbmc/pvr/PVREventlogJob.h]: > **Exception:** Do not indent namespaces to simplify nesting them and wrapping .cpp files in a namespace.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

File: xbmc/pvr/PVRGUIChannelNavigator.cpp [hunk 4]
```diff
@@ -12,15 +12,79 @@
 #include "GUIInfoManager.h"
 #include "ServiceBroker.h"
 #include "guilib/GUIComponent.h"
+#ifdef TARGET_POSIX
+#include "platform/posix/XTimeUtils.h"
+#endif
```
Comment (Rechi) [inline xbmc/pvr/PVRGUIChannelNavigator.cpp]: wrong include order, https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-headers

File: xbmc/pvr/PVRGUIChannelNavigator.cpp [hunk 5]
```diff
@@ -12,15 +12,79 @@
...
+
+  class CPVRChannelInfoTimeoutJob : public CPVRChannelTimeoutJobBase
+  {
+  public:
+    CPVRChannelInfoTimeoutJob(PVR::CPVRGUIChannelNavigator& channelNavigator, int iTimeout)
+    : CPVRChannelTimeoutJobBase(channelNavigator, iTimeout) {}
+    ~CPVRChannelInfoTimeoutJob() override = default;
+    const char* GetType() const override { return "pvr-channel-info-timeout-job"; }
+    void OnTimeout() override { m_channelNavigator.HideInfo(); }
+  };
```
Comment (Rechi) [inline xbmc/pvr/PVRGUIChannelNavigator.cpp]: > **Exception:** Do not indent namespaces to simplify nesting them and wrapping .cpp files in a namespace.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

File: xbmc/pvr/PVRManager.cpp [hunk 6]
```diff
@@ -44,9 +44,62 @@
...
+
+
+  private:
+    void AppendJob(CJob* job);
+
+    CCriticalSection m_critSection;
+    CEvent m_triggerEvent;
+    std::vector<CJob*> m_pendingUpdates;
+    bool m_bStopped = true;
+  };
```
Comment (Rechi) [inline xbmc/pvr/PVRManager.cpp]: > **Exception:** Do not indent namespaces to simplify nesting them and wrapping .cpp files in a namespace.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#32-indentation

**Comment by ksooo** [comment]
@Rechi @phunkyfish all requested/suggested changes are in. Please comment or approve.

**Comment by ksooo** [comment]
jenkins build this pleaes

**Comment by ksooo** [comment]
jenkins build this please

File: xbmc/pvr/PVREventLogJob.h [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+  struct Event
+  {
+    bool m_bNotifyUser;
+    bool m_bError;
+    std::string m_label;
+    std::string m_msg;
+    std::string m_icon;
+
+    Event(bool bNotifyUser, bool bError, const std::string& label, const std::string& msg, const std::string& icon)
+    : m_bNotifyUser(bNotifyUser), m_bError(bError), m_label(label), m_msg(msg), m_icon(icon) {}
```
Comment (Rechi) [inline xbmc/pvr/PVREventLogJob.h]:
```suggestion
      : m_bNotifyUser(bNotifyUser)
      , m_bError(bError)
      , m_label(label)
      , m_msg(msg)
      , m_icon(icon)
    {
    }
```

File: xbmc/pvr/PVRGUIChannelNavigator.cpp [hunk 8]
```diff
@@ -13,14 +13,78 @@
...
+#endif
+
+namespace
+{
+class CPVRChannelTimeoutJobBase : public CJob, public IJobCallback
+{
+public:
+  CPVRChannelTimeoutJobBase() = delete;
+  CPVRChannelTimeoutJobBase(PVR::CPVRGUIChannelNavigator& channelNavigator, int iTimeout)
+  : m_channelNavigator(channelNavigator)
```
Comment (Rechi) [inline xbmc/pvr/PVRGUIChannelNavigator.cpp]:
```suggestion
    : m_channelNavigator(channelNavigator)
```

File: xbmc/pvr/PVRGUIChannelNavigator.cpp [hunk 9]
```diff
@@ -13,14 +13,78 @@
...
+
+private:
+  XbmcThreads::EndTime m_delayTimer;
+};
+
+class CPVRChannelEntryTimeoutJob : public CPVRChannelTimeoutJobBase
+{
+public:
+  CPVRChannelEntryTimeoutJob(PVR::CPVRGUIChannelNavigator& channelNavigator, int iTimeout)
+  : CPVRChannelTimeoutJobBase(channelNavigator, iTimeout) {}
```
Comment (Rechi) [inline xbmc/pvr/PVRGUIChannelNavigator.cpp]:
```suggestion
    : CPVRChannelTimeoutJobBase(channelNavigator, iTimeout)
  {
  }
```

File: xbmc/pvr/PVRGUIChannelNavigator.cpp [hunk 10]
```diff
@@ -13,14 +13,78 @@
...
+  ~CPVRChannelEntryTimeoutJob() override = default;
+  const char* GetType() const override { return "pvr-channel-entry-timeout-job"; }
+  void OnTimeout() override { m_channelNavigator.SwitchToCurrentChannel(); }
+};
+
+class CPVRChannelInfoTimeoutJob : public CPVRChannelTimeoutJobBase
+{
+public:
+  CPVRChannelInfoTimeoutJob(PVR::CPVRGUIChannelNavigator& channelNavigator, int iTimeout)
+  : CPVRChannelTimeoutJobBase(channelNavigator, iTimeout) {}
```
Comment (Rechi) [inline xbmc/pvr/PVRGUIChannelNavigator.cpp]:
```suggestion
    : CPVRChannelTimeoutJobBase(channelNavigator, iTimeout)
  {
  }
```

File: xbmc/pvr/PVRManager.cpp [hunk 11]
```diff
@@ -44,10 +44,63 @@
...
-: m_triggerEvent(false)
+namespace PVR
 {
-}
+template<typename F>
+class CPVRLambdaJob : public CJob
+{
+public:
+  CPVRLambdaJob() = delete;
+  CPVRLambdaJob(const std::string& type, F&& f) : m_type(type), m_f(std::forward<F>(f)) {}
```
Comment (Rechi) [inline xbmc/pvr/PVRManager.cpp]:
```suggestion
  CPVRLambdaJob(const std::string& type, F&& f)
    : m_type(type)
    , m_f(std::forward<F>(f))
  {
  }
```

File: xbmc/pvr/PVRManager.cpp [hunk 12]
```diff
@@ -44,10 +44,63 @@
...
+
+private:
+  std::string m_type;
+  F m_f;
+};
+
+class CPVRManagerJobQueue
+{
+public:
+  CPVRManagerJobQueue() : m_triggerEvent(false) {}
```
Comment (Rechi) [inline xbmc/pvr/PVRManager.cpp]:
```suggestion
  CPVRManagerJobQueue()
    : m_triggerEvent(false)
  {
  }
```

File: xbmc/pvr/PVREventLogJob.h (hunk 7 above)
Comment (ksooo) [reply xbmc/pvr/PVREventLogJob.h]: Nope. This looks so ugly that I will not accept this coding guideline.

File: xbmc/pvr/PVRGUIChannelNavigator.cpp (hunk 8 above)
Comment (ksooo) [reply xbmc/pvr/PVRGUIChannelNavigator.cpp]: No.

File: xbmc/pvr/PVRGUIChannelNavigator.cpp (hunk 9 above)
Comment (ksooo) [reply xbmc/pvr/PVRGUIChannelNavigator.cpp]: Nope.

File: xbmc/pvr/PVRGUIChannelNavigator.cpp (hunk 10 above)
Comment (ksooo) [reply xbmc/pvr/PVRGUIChannelNavigator.cpp]: Never.

File: xbmc/pvr/PVRManager.cpp (hunk 11 above)
Comment (ksooo) [reply xbmc/pvr/PVRManager.cpp]: Once more, sorry no.

File: xbmc/pvr/PVRManager.cpp (hunk 12 above)
Comment (ksooo) [reply xbmc/pvr/PVRManager.cpp]: Nonono. ;-)

**Review by phunkyfish — APPROVED** [review]

**Comment by phunkyfish** [comment]
Looks good 😉

File: xbmc/pvr/PVREventLogJob.h (hunk 7 above)
Comment (Rechi) [reply xbmc/pvr/PVREventLogJob.h]: As I already told you, if you do not like the style submit a PR that changes the docs and clang-format configuration.

**Pull request merged**